File: java-apps/DeviceServices.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class ServiceInfo {
    String name;
    String displayName;
    String status;
    String startType;
    
    public ServiceInfo(String name, String displayName, String status, String startType) {
        this.name = name;
        this.displayName = displayName;
        this.status = status;
        this.startType = startType;
    }
}

public class DeviceServices {
    
    public static List<ServiceInfo> getWindowsServices() {
        List<ServiceInfo> services = new ArrayList<>();
        
        try {
            Process process = Runtime.getRuntime().exec("sc query type= service state= all");
            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            
            String line;
            String serviceName = "";
            String displayName = "";
            String status = "";
            
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                
                if (line.startsWith("SERVICE_NAME:")) {
                    serviceName = line.substring(13).trim();
                } else if (line.startsWith("DISPLAY_NAME:")) {
                    displayName = line.substring(13).trim();
                } else if (line.contains("STATE")) {
                    String[] parts = line.split(":");
                    if (parts.length > 1) {
                        status = parts[1].trim().split("\\s+")[0];
                    }
                    
                    if (!serviceName.isEmpty()) {
                        services.add(new ServiceInfo(serviceName, displayName, status, "N/A"));
                        serviceName = "";
                        displayName = "";
                        status = "";
                    }
                }
            }
            
            reader.close();
            process.waitFor();
            
        } catch (Exception e) {
            System.err.println("Error reading services: " + e.getMessage());
        }
        
        return services;
    }
    
    public static void displayAllServices(List<ServiceInfo> services) {
        System.out.println("\n========================================");
        System.out.println("       WINDOWS DEVICE SERVICES");
        System.out.println("========================================\n");
        System.out.println("Total Services Found: " + services.size() + "\n");
        
        System.out.printf("%-40s %-50s %-15s%n", "Service Name", "Display Name", "Status");
        System.out.println("-".repeat(105));
        
        for (ServiceInfo service : services) {
            System.out.printf("%-40s %-50s %-15s%n", 
                truncate(service.name, 40), 
                truncate(service.displayName, 50), 
                service.status);
        }
    }
    
    public static void searchService(List<ServiceInfo> services, String searchTerm) {
        System.out.println("\nSearching for: " + searchTerm + "\n");
        boolean found = false;
        
        for (ServiceInfo service : services) {
            if (service.name.toLowerCase().contains(searchTerm.toLowerCase()) || 
                service.displayName.toLowerCase().contains(searchTerm.toLowerCase())) {
                System.out.println("Service Name: " + service.name);
                System.out.println("Display Name: " + service.displayName);
                System.out.println("Status: " + service.status);
                System.out.println("-".repeat(50));
                found = true;
            }
        }
        
        if (!found) {
            System.out.println("No services found matching: " + searchTerm);
        }
    }
    
    public static void filterByStatus(List<ServiceInfo> services, String status) {
        System.out.println("\nServices with status: " + status + "\n");
        int count = 0;
        
        for (ServiceInfo service : services) {
            if (service.status.equalsIgnoreCase(status)) {
                System.out.println(service.displayName + " (" + service.name + ")");
                count++;
            }
        }
        
        System.out.println("\nTotal: " + count + " services");
    }
    
    public static String truncate(String str, int length) {
        if (str.length() > length) {
            return str.substring(0, length - 3) + "...";
        }
        return str;
    }
    
    public static void main(String[] args) {
        System.out.println("Reading Windows Device Services...\n");
        
        List<ServiceInfo> services = getWindowsServices();
        
        if (services.isEmpty()) {
            System.out.println("No services found or access denied.");
            System.out.println("Try running as Administrator.");
            return;
        }
        
        Scanner scanner = new Scanner(System.in);
        int choice;
        
        do {
            System.out.println("\n========================================");
            System.out.println("           MENU OPTIONS");
            System.out.println("========================================");
            System.out.println("1. Display All Services");
            System.out.println("2. Search Service");
            System.out.println("3. Filter by Status (RUNNING)");
            System.out.println("4. Filter by Status (STOPPED)");
            System.out.println("5. Refresh Services List");
            System.out.println("6. Exit");
            System.out.print("Enter your choice: ");
            
            choice = scanner.nextInt();
            scanner.nextLine();
            
            switch (choice) {
                case 1:
                    displayAllServices(services);
                    break;
                case 2:
                    System.out.print("Enter service name to search: ");
                    String searchTerm = scanner.nextLine();
                    searchService(services, searchTerm);
                    break;
                case 3:
                    filterByStatus(services, "RUNNING");
                    break;
                case 4:
                    filterByStatus(services, "STOPPED");
                    break;
                case 5:
                    System.out.println("Refreshing services list...");
                    services = getWindowsServices();
                    System.out.println("Services refreshed!");
                    break;
                case 6:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Invalid choice!");
            }
        } while (choice != 6);
        
        scanner.close();
    }
}

File: java-apps/device_services.cpp
#include <windows.h>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

struct ServiceInfo {
    string name;
    string displayName;
    string status;
    string startType;
};

string getServiceStatus(DWORD status) {
    switch (status) {
        case SERVICE_STOPPED: return "STOPPED";
        case SERVICE_START_PENDING: return "START_PENDING";
        case SERVICE_STOP_PENDING: return "STOP_PENDING";
        case SERVICE_RUNNING: return "RUNNING";
        case SERVICE_CONTINUE_PENDING: return "CONTINUE_PENDING";
        case SERVICE_PAUSE_PENDING: return "PAUSE_PENDING";
        case SERVICE_PAUSED: return "PAUSED";
        default: return "UNKNOWN";
    }
}

string getStartType(DWORD startType) {
    switch (startType) {
        case SERVICE_AUTO_START: return "AUTO";
        case SERVICE_BOOT_START: return "BOOT";
        case SERVICE_DEMAND_START: return "MANUAL";
        case SERVICE_DISABLED: return "DISABLED";
        case SERVICE_SYSTEM_START: return "SYSTEM";
        default: return "UNKNOWN";
    }
}

vector<ServiceInfo> getDeviceServices() {
    vector<ServiceInfo> services;
    SC_HANDLE scManager = OpenSCManager(NULL, NULL, SC_MANAGER_ENUMERATE_SERVICE);
    
    if (scManager == NULL) {
        cerr << "Failed to open Service Control Manager. Error: " << GetLastError() << endl;
        return services;
    }
    
    DWORD bytesNeeded = 0;
    DWORD servicesReturned = 0;
    DWORD resumeHandle = 0;
    
    EnumServicesStatusEx(
        scManager,
        SC_ENUM_PROCESS_INFO,
        SERVICE_WIN32,
        SERVICE_STATE_ALL,
        NULL,
        0,
        &bytesNeeded,
        &servicesReturned,
        &resumeHandle,
        NULL
    );
    
    LPBYTE buffer = new BYTE[bytesNeeded];
    ENUM_SERVICE_STATUS_PROCESS* serviceStatus = (ENUM_SERVICE_STATUS_PROCESS*)buffer;
    
    if (EnumServicesStatusEx(
        scManager,
        SC_ENUM_PROCESS_INFO,
        SERVICE_WIN32,
        SERVICE_STATE_ALL,
        buffer,
        bytesNeeded,
        &bytesNeeded,
        &servicesReturned,
        &resumeHandle,
        NULL
    )) {
        for (DWORD i = 0; i < servicesReturned; i++) {
            SC_HANDLE service = OpenService(scManager, serviceStatus[i].lpServiceName, SERVICE_QUERY_CONFIG);
            
            if (service != NULL) {
                DWORD configBytesNeeded = 0;
                QueryServiceConfig(service, NULL, 0, &configBytesNeeded);
                
                LPQUERY_SERVICE_CONFIG serviceConfig = (LPQUERY_SERVICE_CONFIG)new BYTE[configBytesNeeded];
                
                if (QueryServiceConfig(service, serviceConfig, configBytesNeeded, &configBytesNeeded)) {
                    ServiceInfo info;
                    info.name = serviceStatus[i].lpServiceName;
                    info.displayName = serviceStatus[i].lpDisplayName;
                    info.status = getServiceStatus(serviceStatus[i].ServiceStatusProcess.dwCurrentState);
                    info.startType = getStartType(serviceConfig->dwStartType);
                    
                    services.push_back(info);
                }
                
                delete[] serviceConfig;
                CloseServiceHandle(service);
            }
        }
    }
    
    delete[] buffer;
    CloseServiceHandle(scManager);
    
    return services;
}

void displayServices(const vector<ServiceInfo>& services) {
    cout << "\n========================================" << endl;
    cout << "       WINDOWS DEVICE SERVICES" << endl;
    cout << "========================================\n" << endl;
    cout << "Total Services Found: " << services.size() << "\n" << endl;
    
    cout.width(40); cout << left << "Service Name";
    cout.width(50); cout << left << "Display Name";
    cout.width(15); cout << left << "Status";
    cout.width(10); cout << left << "Start Type" << endl;
    cout << string(115, '-') << endl;
    
    for (const auto& service : services) {
        cout.width(40); cout << left << service.name;
        cout.width(50); cout << left << service.displayName;
        cout.width(15); cout << left << service.status;
        cout.width(10); cout << left << service.startType << endl;
    }
}

void searchService(const vector<ServiceInfo>& services, const string& searchTerm) {
    cout << "\nSearching for: " << searchTerm << "\n" << endl;
    bool found = false;
    
    for (const auto& service : services) {
        if (service.name.find(searchTerm) != string::npos || 
            service.displayName.find(searchTerm) != string::npos) {
            cout << "Service Name: " << service.name << endl;
            cout << "Display Name: " << service.displayName << endl;
            cout << "Status: " << service.status << endl;
            cout << "Start Type: " << service.startType << endl;
            cout << string(50, '-') << endl;
            found = true;
        }
    }
    
    if (!found) {
        cout << "No services found matching: " << searchTerm << endl;
    }
}

void filterByStatus(const vector<ServiceInfo>& services, const string& status) {
    cout << "\nServices with status: " << status << "\n" << endl;
    int count = 0;
    
    for (const auto& service : services) {
        if (service.status == status) {
            cout << service.displayName << " (" << service.name << ")" << endl;
            count++;
        }
    }
    
    cout << "\nTotal: " << count << " services" << endl;
}

int main() {
    cout << "Reading Windows Device Services..." << endl;
    
    vector<ServiceInfo> services = getDeviceServices();
    
    if (services.empty()) {
        cout << "No services found or access denied." << endl;
        return 1;
    }
    
    int choice;
    do {
        cout << "\n========================================" << endl;
        cout << "           MENU OPTIONS" << endl;
        cout << "========================================" << endl;
        cout << "1. Display All Services" << endl;
        cout << "2. Search Service" << endl;
        cout << "3. Filter by Status (RUNNING)" << endl;
        cout << "4. Filter by Status (STOPPED)" << endl;
        cout << "5. Exit" << endl;
        cout << "Enter your choice: ";
        cin >> choice;
        cin.ignore();
        
        switch (choice) {
            case 1:
                displayServices(services);
                break;
            case 2: {
                string searchTerm;
                cout << "Enter service name to search: ";
                getline(cin, searchTerm);
                searchService(services, searchTerm);
                break;
            }
            case 3:
                filterByStatus(services, "RUNNING");
                break;
            case 4:
                filterByStatus(services, "STOPPED");
                break;
            case 5:
                cout << "Exiting..." << endl;
                break;
            default:
                cout << "Invalid choice!" << endl;
        }
    } while (choice != 5);
    
    return 0;
}
